resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java

```diff
@@ -64,7 +64,7 @@ public void operationComplete(Future<List<InetAddress>> future) throws Exception
                     if (numAddresses > 0) {
                         // if there are multiple addresses: we shall pick one by one
                         // to support the round robin distribution
-                        promise.setSuccess(inetAddresses.get(index.getAndIncrement() % numAddresses));
+                        promise.setSuccess(inetAddresses.get(toIndex(counter.getAndIncrement(), numAddresses)));
```

Comment:
`Math.abs` changes direction as stated, that's the problem for neg counter and `%` operation, the current proposed implementation is the best you can get for non-power of two sizes.

For power of two which is not the case you can always do it faster by shifting bits but again, not the case here, and for a perfect rotation you would have to use a `CAS` operation which is at least 5 times slower the last time I benchmark it using JDK 8u40+